dlib/dnn/tensor_tools.h

```diff
@@ -947,6 +947,24 @@ namespace dlib { namespace tt
                   this gradient to filters_gradient.
         !*/
 
+        void operator() (
+            resizable_tensor& output,
+            const tensor& data,
+            const tensor& filters
+       )
+       {
+          impl.forward_without_setup(output,data,filters);
+       }
+
+       void setup(
```

Comment:
This whole PR looks great, except for this part where there are these two undocumented functions.  They also behave in a somewhat opposite way compared to the others, which makes the class interface as a whole not cohesive and hard to understand.  I get what it's doing, but interface complexity is one of the roots of software bugs, and so should be avoided.

Really the interface for this class should either require some kind of setup before use for all the functions uniformly, or not require setup at all, as it previously did.  Not a mixture of both, and especially not one that isn't documented in any way.  

So what about this.  There can be a setup that takes only the 4 stride and padding arguments.  Calling that function will be required before any other function is executed.  Then the rest of the functions can do an internal allocation, initialization, or whatever, if they need it, to make everything work.  This would mean removing the other operator() that takes stride and padding.  It seems like that would cover everything you want to do but provide for a uniform and documentable interface.  Does that sound right or am I missing something?